test/specs/lib/FocusZone-test.tsx

```diff
@@ -0,0 +1,1451 @@
+          <button className="a">a</button>
+          <button className="b">b</button>
+          <button className="c">c</button>
+        </FocusZone>
+      </div>,
+    )
+
+    const focusZone = ReactDOM.findDOMNode(component as React.ReactInstance)!!.firstChild as Element
+
+    const buttonA = focusZone.querySelector('.a') as HTMLElement
+    const buttonB = focusZone.querySelector('.b') as HTMLElement
+    const buttonC = focusZone.querySelector('.c') as HTMLElement
+
+    // Assign bounding locations to buttons.
+    setupElement(buttonA, {
+      clientRect: {
+        top: 0,
+        bottom: 30,
+        left: 0,
+        right: 100,
+      },
+    })
+
+    setupElement(buttonB, {
+      clientRect: {
+        top: 30,
+        bottom: 60,
+        left: 0,
+        right: 100,
+      },
+    })
+
+    setupElement(buttonC, {
+      clientRect: {
+        top: 60,
+        bottom: 90,
+        left: 0,
+        right: 100,
+      },
+    })
+
+    // Focus the first button.
+    ReactTestUtils.Simulate.focus(buttonA)
+    expect(lastFocusedElement).toBe(buttonA)
+
+    // Pressing down should go to b.
+    ReactTestUtils.Simulate.keyDown(focusZone, { which: KeyCodes.down })
+    expect(lastFocusedElement).toBe(buttonB)
+
+    // Pressing down should go to c.
+    ReactTestUtils.Simulate.keyDown(focusZone, { which: KeyCodes.down })
+    expect(lastFocusedElement).toBe(buttonC)
+
+    // Pressing down should stay on c.
+    ReactTestUtils.Simulate.keyDown(focusZone, { which: KeyCodes.down })
+    expect(lastFocusedElement).toBe(buttonC)
+
+    // Pressing up should go to b.
+    ReactTestUtils.Simulate.keyDown(focusZone, { which: KeyCodes.up })
+    expect(lastFocusedElement).toBe(buttonB)
+
+    // Pressing up should go to a.
+    ReactTestUtils.Simulate.keyDown(focusZone, { which: KeyCodes.up })
+    expect(lastFocusedElement).toBe(buttonA)
+
+    // Pressing up should stay on a.
+    ReactTestUtils.Simulate.keyDown(focusZone, { which: KeyCodes.up })
+    expect(lastFocusedElement).toBe(buttonA)
+
+    // Click on c to focus it.
+    ReactTestUtils.Simulate.focus(buttonC)
+    expect(lastFocusedElement).toBe(buttonC)
+
+    // Pressing up should move to b.
+    ReactTestUtils.Simulate.keyDown(focusZone, { which: KeyCodes.up })
+    expect(lastFocusedElement).toBe(buttonB)
+
+    // Test that pressing horizontal buttons don't move focus.
+    ReactTestUtils.Simulate.keyDown(focusZone, { which: KeyCodes.left })
+    expect(lastFocusedElement).toBe(buttonB)
+
+    ReactTestUtils.Simulate.keyDown(focusZone, { which: KeyCodes.right })
+    expect(lastFocusedElement).toBe(buttonB)
+
+    // Press home should go to the first target.
+    ReactTestUtils.Simulate.keyDown(focusZone, { which: KeyCodes.home })
+    expect(lastFocusedElement).toBe(buttonA)
+
+    // // Press end should go to the last target.
+    ReactTestUtils.Simulate.keyDown(focusZone, { which: KeyCodes.end })
+    expect(lastFocusedElement).toBe(buttonC)
+  })
```

Comment:
This is a very large test that would be tough to debug in a failure.  Is it possible to break these down to single purposed assertions?  Things like "test('home goes to first target')`.

It would make the pass/fail messages more helpful, tests easier to debug, and help developers understand what the component does just by reading tests.

Reply:
When a test like this fails, the following info gets also printed to console:

```
      122 |     // Pressing up should stay on a.
      123 |     ReactTestUtils.Simulate.keyDown(focusZone, { which: keyboardKey.ArrowUp })
    > 124 |     expect(lastFocusedElement).toBe(buttonB)
```

This way it is extremely easy to see what exactly failed. I do not think it's worth separating everything into a new test especially when these tests are actually ordered and depend on the previous keyboard navigation that occured before.

It might be worth it considering adding some easy way to add custom expect messages to Stardust like it is possible in the Jasmine framework, e.g., using something like [jest-expect-message](https://github.com/mattphillips/jest-expect-message).